src/cert/mod.rs

```diff
@@ -109,6 +114,168 @@ pub async fn upload_new_cert_ref(
+        "Expired".to_string()
+    } else if duration.num_hours() >= 24 {
+        format!("Expires in {} days", duration.num_days())
+    } else {
+        format!("Expires in {} hours", duration.num_hours())
+    };
+
+    Ok(formatted_duration)
+}
+
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+struct CertWithFormattedString {
+    cert: CageSigningCert,
+    formatted: String,
+    locked: bool,
+}
+
+impl CertWithFormattedString {
+    fn new(cert: &CageSigningCert, locked: bool) -> Self {
+        Self {
+            formatted: format_cert_for_multi_select(cert),
+            cert: cert.clone(),
+            locked,
+        }
+    }
+}
+
+async fn get_certs_for_selection(
+    cage_api: api::cage::CagesClient,
+    cage_uuid: &str,
+) -> Result<Vec<CertWithFormattedString>, CertError> {
+    let available_certs = match cage_api.get_signing_certs().await {
+        Ok(res) => res.certs,
+        Err(e) => {
+            log::error!("Error getting cage signing cert refs — {:?}", e);
+            return Err(CertError::ApiError(e));
+        }
+    };
+
+    let locked_certs = match cage_api.get_cage_locked_signing_certs(cage_uuid).await {
+        Ok(certs) => certs,
+        Err(e) => {
+            log::error!("Error getting cage signing cert — {:?}", e);
+            return Err(CertError::ApiError(e));
+        }
+    };
+
+    let locked_cert_uuids = locked_certs
+        .iter()
+        .map(|cert| cert.uuid())
+        .collect::<Vec<&str>>();
+
+    let available_stripped: Vec<&CageSigningCert> = available_certs
+        .iter()
+        .filter(|cert| !locked_cert_uuids.contains(&cert.uuid()))
+        .collect();
+
+    let available_formatted = available_stripped
+        .iter()
+        .map(|cert| CertWithFormattedString::new(cert, false))
+        .collect::<Vec<CertWithFormattedString>>();
+    let locked_formatted = locked_certs
+        .iter()
+        .map(|cert| CertWithFormattedString::new(cert, true))
+        .collect::<Vec<CertWithFormattedString>>();
+
+    let mut all_formatted: Vec<CertWithFormattedString> = vec![];
+
+    all_formatted.extend(available_formatted);
+    all_formatted.extend(locked_formatted);
+
+    Ok(all_formatted)
+}
+
+fn sort_certs_by_expiry(
+    mut certs: Vec<CertWithFormattedString>,
+) -> Result<Vec<CertWithFormattedString>, CertError> {
+    certs.sort_by_key(|cert| cert.cert.not_after().unwrap_or("".to_string()));
+    Ok(certs)
+}
+
+pub async fn lock_cage_to_certs(api_key: &str, cage_uuid: &str) -> Result<(), CertError> {
+    let cage_api = api::cage::CagesClient::new(AuthMode::ApiKey(api_key.to_string()));
+
+    let certs_for_select = get_certs_for_selection(cage_api.clone(), cage_uuid).await?;
+
+    let sorted_certs_for_select = sort_certs_by_expiry(certs_for_select)?;
+
+    let chosen: Vec<usize> = MultiSelect::new()
+        .with_prompt("Select Certs To Lock Cage To. Press Space To Select, Enter To Confirm.\n Cert Name | PCR8 (Hash of cert) | Cert Expiry ")
```

Comment:
surprised there isn't headers support for dialoguer!

Reply:
Yeah, I thought it was strange